File: другое/тинькофф.аисд/2/G.ts
const n = 7;
const lst = [
  ["+", "1"],
  ["+", "2"],
  ["-"],
  ["+", "3"],
  ["+", "4"],
  ["-"],
  ["-"],
];

const left: number[] = [];
const right: number[] = [];

for (const item of lst) {
  if (item[0] === "+") {
    right.push(Number(item[1]));
    if (left.length < right.length) {
      left.push(right.shift()!);
    }
  } else if (item[0] === "*") {
    if (left.length === right.length) {
      left.push(Number(item[1]));
    } else {
      right.push(Number(item[1]));
    }
  } else {
    console.log(left.shift()!);

    const b = right.shift();
    if (b) left.push(b);
  }
}

File: другое/тинькофф.аисд/2/G.py
from collections import deque

n = int(input())

left = deque()
right = deque()

for _ in range(n):
    item = input().split()
    if item[0] == "+":
        right.append(int(item[1]))
        if len(left) < len(right):
            left.append(right.popleft())
    elif item[0] == "*":
        if len(left) == len(right):
            left.append(int(item[1]))
        else:
            right.appendleft(int(item[1]))
    else:
        print(left.popleft())
        if len(left) < len(right) and len(right) > 0:
            left.append(right.popleft())
